File: 力扣/学习计划/剑指 Offer I/016_从上到下打印二叉树 III/main.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

import common.TreeNode;

/**
 * https://leetcode-cn.com/problems/cong-shang-dao-xia-da-yin-er-cha-shu-iii-lcof
 */
class Solution {
    /**
     * 双栈
     */
    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null) return new ArrayList<>();

        List<List<Integer>> resList = new ArrayList<>();
        Stack<TreeNode> stack1 = new Stack<>();
        Stack<TreeNode> stack2 = new Stack<>();
        stack1.add(root);

        boolean flag = true;
        while (!(stack1.isEmpty() && stack2.isEmpty())) {
            List<Integer> tmpList = new ArrayList<>();
            if (flag) {
                for (int i = stack1.size(); i > 0; i--) {
                    TreeNode node = stack1.pop();
                    tmpList.add(node.val);
                    if (node.left != null)
                        stack2.add(node.left);
                    if (node.right != null)
                        stack2.add(node.right);
                }
            } else {
                for (int i = stack2.size(); i > 0; i--) {
                    TreeNode node = stack2.pop();
                    tmpList.add(node.val);
                    if (node.right != null)
                        stack1.add(node.right);
                    if (node.left != null)
                        stack1.add(node.left);
                }

            }
            resList.add(tmpList);
            flag = !flag;
        }
        return resList;
    }
}

File: 力扣/学习计划/剑指 Offer I/016_从上到下打印二叉树 III/main.go
package main

func levelOrder(root *TreeNode) [][]int {
	var res = make([][]int, 0)
	if root == nil {
		return res
	}

	queue := []*TreeNode{root}

	// 奇偶顺序控制
	flag := false
	for len(queue) > 0 {
		length := len(queue)
		tmp := make([]int, length)

		for i := 0; i < length; i++ {
			node := queue[0]
			queue = queue[1:]
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
			if flag {
				tmp[length-i-1] = node.Val
			} else {
				tmp[i] = node.Val
			}
		}
		res = append(res, tmp)
		flag = !flag
	}
	return res
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}
